### PR TITLE
Kubebuilder v2 Migration

### LINKED ISSUE #502: Migrate all Crossplane controllers to use kubebuilder 2.0.0 and controller-runtime -0.2.0
https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v2.0.0-alpha.2
https://github.com/kubernetes-sigs/controller-runtime/releases/tag/v0.2.0-beta.1

These new releases will likely include some breaking changes. It may be worth assessing how different they are and how soon we would like to upgrade before we invest significantly in standardising our controller patterns.

### PR DESCRIPTION
**Pull request by hasheddan**
This PR replaces #665, which had become cluttered with comments and commits.

### Description of your changes
This PR represents @muvaf work (along with myself and @jbw976) to migrate Crossplane to the Kubebuilder v2 [structure](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/simplified-scaffolding.md). It also involves migration to [controller-runtime v0.2](https://github.com/kubernetes-sigs/controller-runtime) and a temporary fork of [controller-tools v0.2](https://github.com/muvaf/controller-tools)(more on this below). The primary motivations for this migration are as follows:

* Consistency with the Kubebuilder framework, which will likely be the tool that future stack developers are familiar with
* Enables a seamless break-out of major cloud-provider (AWS, GCP, Azure) apis, clients, and controllers intro their own individual stacks
* Ability to continue to utilize the functionality provided by the `controller-runtime` and `controller-tools` libraries

### Additional Details
The commits in this PR have been rebased to show the categories of changes that are implemented. The most prominent changes that can be observed are as follows:

* The `apis` for each cloud provider have been brought to the top-level directory
* The comment markers used by `kubebuilder` and `controller-tools`  to generate `DeepCopy` functions and CRD's have been updated to [new style](https://book.kubebuilder.io/reference/markers/crd.html)
* The `AddToManagerFuncs` pattern has been replaced by the `SetupWithManager` pattern, which will allow us to [pass configuration details](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/simplified-scaffolding.md#dependency-passing-woes) down to each controller at runtime in the future
* We are taking a temporary dependency on @muvaf [fork](https://github.com/muvaf/controller-tools) of `controller-tools` in expectation that the change tracked by issue [controller-tools/#302](https://github.com/kubernetes-sigs/controller-tools/issues/302) will be resolved in the future
* The `build` submodule has been updated to correct any `goimports` errors that are introduced as a result of `make generate`, which we experienced after migrating to `controller-tools` to v0.2 

Fixes #502 

### TODO
There are a few minor issues that will need to be addressed before merging this PR. They will not affect functionality, but are currently causing unit test failures.
- [x] Fix `BucketPolicyOnly` error in `make manifests`
- [x] Fix `localhost:8080` unit test error
- [x] Reassess new dependency rules in `Gopkg.toml` 

### Checklist
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [x] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -42,12 +42,29 @@ required = [
...
+  version="v0.2.0-beta.5"
 
 [[override]]
   name="sigs.k8s.io/controller-tools"
-  source = "https://github.com/crossplaneio/controller-tools.git"
-  revision = "9acc926ef7ce7372609b34fbafe397158e6442ef"
+  source = "https://github.com/muvaf/controller-tools.git"
+  revision = "301a879da11105676f319f9080278a54b956d07f"
+
+# temp stanzas
```
Comment (jbw976) [inline Gopkg.toml]: we should take another pass through these dependency rules to see if any can be removed.  I have a feeling that some were needed while we were still getting through the migration, but they might not all be needed at a steady state.

https://golang.github.io/dep/docs/Gopkg.toml.html

**Comment by jbw976** [comment]
The very first Jenkins build hit the intermittent unit test failure of port 8080 already in use 🤕 

https://jenkinsci.upbound.io/blue/organizations/jenkins/crossplane%2Fbuild/detail/PR-675/1/pipeline:
```
[2019-08-13T23:38:55.934Z] === RUN   TestReconcile
[2019-08-13T23:38:55.934Z] --- FAIL: TestReconcile (0.06s)
[2019-08-13T23:38:55.934Z]     sqlserver_test.go:159: 
[2019-08-13T23:38:55.934Z]         Unexpected error:
[2019-08-13T23:38:55.934Z]             <*errors.errorString | 0xc000739b20>: {
[2019-08-13T23:38:55.934Z]                 s: "error listening on :8080: listen tcp :8080: bind: address already in use",
[2019-08-13T23:38:55.934Z]             }
[2019-08-13T23:38:55.934Z]             error listening on :8080: listen tcp :8080: bind: address already in use
[2019-08-13T23:38:55.934Z]         occurred
[2019-08-13T23:38:55.934Z] FAIL
```

**Comment by displague** [comment]
@jbw976 I ran into this while trying to debug crossplane on this branch and temporarily patched the metrics server out:
```patch
diff --git a/cmd/crossplane/main.go b/cmd/crossplane/main.go
index ac76294..fa55c41 100644
--- a/cmd/crossplane/main.go
+++ b/cmd/crossplane/main.go
@@ -115,7 +115,7 @@ func main() {
 	log.Info("Sync period", "duration", syncPeriod.String())
 
 	// Create a new Cmd to provide shared dependencies and start components
-	mgr, err := manager.New(cfg, manager.Options{SyncPeriod: syncPeriod})
+	mgr, err := manager.New(cfg, manager.Options{SyncPeriod: syncPeriod, MetricsBindAddress: "0"})
 	if err != nil {
 		kingpin.FatalIfError(err, "Cannot create manager")
 	}
```

Maybe we should make this an environmental setting?

Why is this only becoming a problem now?

**Comment by hasheddan** [comment]
I have corrected the port `8080` error in a pending commit.

File: apis/cache/v1alpha1/doc.go [hunk 2]
```diff
@@ -15,5 +15,4 @@ limitations under the License.
 */
 
 // Package v1alpha1 contains API Schema definitions for the cache v1alpha1 API group
-// +k8s:deepcopy-gen=package,register
 package v1alpha1
```
Comment (soorena776) [inline apis/cache/v1alpha1/doc.go]: do we still need to keep these `doc.go` files?

File: Gopkg.toml (hunk 1 above)
Comment (soorena776) [reply Gopkg.toml]: Is there a reason we are keeping this file?
If you run `go mod init github.com/crossplaneio/crossplane` in this folder it automatically converts `Gopkg.toml` to `go.mod`, and you can get rid of this file.

File: apis/cache/v1alpha1/register.go [hunk 3]
```diff
@@ -13,6 +13,9 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+// +kubebuilder:object:generate=true
+// +groupName=cache.crossplane.io
+// +versionName=v1alpha1
 
 package v1alpha1
```
Comment (soorena776) [inline apis/cache/v1alpha1/register.go]: should we still keep this file?

Comment (hasheddan) [reply apis/cache/v1alpha1/register.go]: Could you elaborate as to why this file should be removed?

**Comment by hasheddan** [comment]
`BucketPolicyOnly` issue has been fixed in a pending commit.

File: apis/extensions/v1alpha1/types.go [hunk 4]
```diff
@@ -114,8 +114,8 @@ type ExtensionSpec struct {
 
 // ExtensionStatus defines the observed state of Extension
 type ExtensionStatus struct {
-	corev1alpha1.ConditionedStatus
-	ControllerRef *corev1.ObjectReference `json:"controllerRef,omitempty"`
+	v1alpha1.ConditionedStatus `json:"conditionedStatus,omitempty"`
+	ControllerRef              *corev1.ObjectReference `json:"controllerRef,omitempty"`
```
Comment (soorena776) [inline apis/extensions/v1alpha1/types.go]: nit: space vs tab

File: apis/workload/v1alpha1/types.go [hunk 5]
```diff
@@ -55,6 +55,9 @@ type KubernetesApplicationSpec struct {
 	// KubernetesApplicationResources.
 	ResourceSelector *metav1.LabelSelector `json:"resourceSelector"`
 
+	// TODO(muvaf): Only MatchLabels field of LabelSelector is used. Incorporate
```
Comment (soorena776) [inline apis/workload/v1alpha1/types.go]: Are we going to have TODOs in the code? Shouldn't we just open an issue instead?

File: Gopkg.toml (hunk 1 above)
Comment (hasheddan) [reply Gopkg.toml]: @soorena776 while I support the movement to go modules, I think that should be addressed in a separate dedicated PR.

File: apis/workload/v1alpha1/types.go [hunk 6]
```diff
@@ -79,7 +82,7 @@ type KubernetesApplicationResourceTemplate struct {
 // KubernetesApplicationStatus represents the status of a Kubernetes
 // application.
 type KubernetesApplicationStatus struct {
-	corev1alpha1.ConditionedStatus
+	corev1alpha1.ConditionedStatus `json:"conditionedStatus,omitempty"`
```
Comment (soorena776) [inline apis/workload/v1alpha1/types.go]: is tagging the types a new requirement in V2, or did we just miss it?

File: aws/apis/aws.go [hunk 7]
```diff
@@ -14,26 +14,29 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
```
Comment (soorena776) [inline aws/apis/aws.go]: why this filename change? gcp to aws?

File: azure/apis/azure.go [hunk 8]
```diff
@@ -14,27 +14,30 @@ See the License for the specific language governing permissions and
 limitations under the License.
```
Comment (soorena776) [inline azure/apis/azure.go]: same weird filename change here

File: azure/apis/database/v1alpha1/sql_types.go [hunk 9]
```diff
@@ -33,6 +34,8 @@ const (
 	OperationCreateFirewallRules = "createFirewallRules"
 )
 
+// +kubebuilder:object:generate=false
```
Comment (soorena776) [inline azure/apis/database/v1alpha1/sql_types.go]: is there a reason for `generate=false`?

File: aws/apis/aws.go (hunk 7 above)
Comment (hasheddan) [reply aws/apis/aws.go]: Git internally does not really care about renames, it just tracks content. It uses a `similarity index` to guess if a file is a rename. This can be adjusted in a local git config if desired. I am guessing GitHub is detecting a <50% change in content here so is assuming it is a rename. More context [here](https://git-scm.com/docs/git-diff-index#Documentation/git-diff-index.txt---find-renamesltngt).

File: azure/apis/azure.go (hunk 8 above)
Comment (hasheddan) [reply azure/apis/azure.go]: See response to previous comment.

File: apis/workload/v1alpha1/types.go (hunk 6 above)
Comment (hasheddan) [reply apis/workload/v1alpha1/types.go]: Tagging is required by `controller-tools` in V2.

File: cluster/charts/crossplane/crds/aws/aws.crossplane.io_providers.yaml [hunk 10]
```diff
@@ -1,9 +1,9 @@
+
```
Comment (soorena776) [inline cluster/charts/crossplane/crds/aws/aws.crossplane.io_providers.yaml]: Is this CRD autogenerated? If not, what are the purpose of these changes?

File: azure/apis/database/v1alpha1/sql_types.go (hunk 9 above)
Comment (hasheddan) [reply azure/apis/database/v1alpha1/sql_types.go]: Yes. Without the tag it will attempt to generate `DeepCopy` functions for the `SQLServer` interface, which we do not want.

File: cluster/charts/crossplane/crds/aws/aws.crossplane.io_providers.yaml (hunk 10 above)
Comment (hasheddan) [reply cluster/charts/crossplane/crds/aws/aws.crossplane.io_providers.yaml]: Yes. All CRD's are autogenerated with `make manifests`.

File: cluster/local/minikube.sh [hunk 11]
```diff
@@ -67,7 +67,6 @@ case "${1:-}" in
     ${HELM} init --service-account tiller
     kubectl -n kube-system rollout status deploy/tiller-deploy
     kubectl -n kube-system rollout status deploy/nginx-ingress-controller
-    kubectl -n kube-system rollout status deploy/default-http-backend
```
Comment (soorena776) [inline cluster/local/minikube.sh]: what is the reason for deleting this?

File: gcp/apis/gcp.go [hunk 12]
```diff
@@ -14,23 +14,26 @@ See the License for the specific language governing permissions and
 limitations under the License.
```
Comment (soorena776) [inline gcp/apis/gcp.go]: same filename change

File: pkg/clients/aws/aws.go [hunk 13]
```diff
@@ -23,12 +23,12 @@ import (
...
 	"k8s.io/client-go/kubernetes"
 
-	"github.com/crossplaneio/crossplane/pkg/apis/aws/v1alpha1"
+	"github.com/crossplaneio/crossplane/aws/apis/v1alpha1"
 	"github.com/crossplaneio/crossplane/pkg/util"
 )
 
 // DefaultSection for INI files.
-const DefaultSection = ini.DEFAULT_SECTION
+const DefaultSection = ini.DefaultSection
```
Comment (soorena776) [inline pkg/clients/aws/aws.go]: is this related to kubebuilder v2?

File: cluster/local/minikube.sh (hunk 11 above)
Comment (hasheddan) [reply cluster/local/minikube.sh]: I believe the `default-http-backend` is not needed due to the installation of `nignx-ingress-controller`.

File: pkg/clients/gcp/cloudsql/instance.go [hunk 14]
```diff
@@ -46,7 +48,7 @@ var _ InstanceService = &InstanceClient{}
 
 // NewInstanceClient creates a new instance of an InstanceClient
 func NewInstanceClient(ctx context.Context, creds *google.Credentials) (*InstanceClient, error) {
-	service, err := sqladmin.New(oauth2.NewClient(ctx, creds.TokenSource))
+	service, err := sqladmin.NewService(ctx, option.WithHTTPClient(oauth2.NewClient(ctx, creds.TokenSource)))
```
Comment (soorena776) [inline pkg/clients/gcp/cloudsql/instance.go]: is this related to kubebuilder v2?

File: apis/cache/v1alpha1/register.go (hunk 3 above)
Comment (soorena776) [reply apis/cache/v1alpha1/register.go]: I see it now, these annotations are kept separately for autogeneration

File: aws/apis/aws.go (hunk 7 above)
Comment (soorena776) [reply aws/apis/aws.go]: as long as we are not mixing providers unintentionally! :)

File: cluster/charts/crossplane/crds/aws/aws.crossplane.io_providers.yaml (hunk 10 above)
Comment (soorena776) [reply cluster/charts/crossplane/crds/aws/aws.crossplane.io_providers.yaml]: that's what I thought, but there is a TODO in the yaml which made me wonder

**Comment by muvaf** [comment]
@hasheddan thanks for having this PR in a cleaner shape that is easy to digest.

While we pushed to the same branch for practical reasons, I think we should try to keep the git history as if it's gone through a proper PR process. So, I think it'd be better if we squash only feature commits that belong to the same author, otherwise commits could stay separate, which is what you'd get when you open a PR against another branch.

File: pkg/controller/gcp/gcp.go [hunk 15]
```diff
@@ -17,38 +17,54 @@ limitations under the License.
...
-}
+// Controllers passes down config and adds individual controllers to the manager.
+type Controllers struct{}
 
-// AddToManagerFuncs is a list of functions to add all Controllers to the Manager
-var AddToManagerFuncs []func(manager.Manager) error
+// SetupWithManager adds all GCP controllers to the manager.
+func (c *Controllers) SetupWithManager(mgr ctrl.Manager) error {
+	if err := (&cache.CloudMemorystoreInstanceClaimController{}).SetupWithManager(mgr); err != nil {
+		return err
```
Comment (suskin) [inline pkg/controller/gcp/gcp.go]: Did we have any particular reason to fail fast for these errors? As a developer I usually find it to be a better experience if I can get back a list of all the errors at once, instead of having the first error hide other ones. In this case, that'd mean collecting all the `err`s and returning them all together as a group. Or aggregated under a single error.

File: pkg/extensions/unpack_test.go [hunk 16]
```diff
@@ -275,7 +275,8 @@ spec:
       - delete
   title: Sample Crossplane Extension
   version: 0.0.1
-status: {}
+status:
+  conditionedStatus: {}
```
Comment (suskin) [inline pkg/extensions/unpack_test.go]: This is probably a dumb question, but what is `conditionedStatus`?

File: pkg/logging/logging.go [hunk 17]
```diff
@@ -29,7 +29,7 @@ const (
 )
 
 var (
-	logger = runtimelog.NewDelegatingLogger(runtimelog.NullLogger{})
+	logger = runtimelog.Log
```
Comment (suskin) [inline pkg/logging/logging.go]: What is the benefit of this change? I don't fully understand it in the context of the migration yet

File: pkg/resource/deprecateddefaultclass_test.go [hunk 18]
```diff
@@ -26,8 +26,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
-	"github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
-	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	corev1alpha1 "github.com/crossplaneio/crossplane/apis/core/v1alpha1"
```
Comment (suskin) [inline pkg/resource/deprecateddefaultclass_test.go]: Nice consolidation

File: aws/apis/aws.go [hunk 19]
```diff
@@ -14,26 +14,29 @@ See the License for the specific language governing permissions and
...
 
-// Package gcp contains Kubernetes API for GCP cloud provider.
-package gcp
+// Generate deepcopy for apis
```
[the remark this answers is not in the document]
Comment (hasheddan) [reply aws/apis/aws.go]: See response to @soorena776 comments.

File: aws/apis/compute/v1alpha1/types.go [hunk 20]
```diff
@@ -183,7 +183,7 @@ type WorkerNodesSpec struct {
...
+	// +kubebuilder:validation:Enum=t2.small;t2.medium;t2.large;t2.xlarge;t2.2xlarge;t3.nano;t3.micro;t3.small;t3.medium;t3.large;t3.xlarge;t3.2xlarge;m3.medium;m3.large;m3.xlarge;m3.2xlarge;m4.large;m4.xlarge;m4.2xlarge;m4.4xlarge;m4.10xlarge;m5.large;m5.xlarge;m5.2xlarge;m5.4xlarge;m5.12xlarge;m5.24xlarge;c4.large;c4.xlarge;c4.2xlarge;c4.4xlarge;c4.8xlarge;c5.large;c5.xlarge;c5.2xlarge;c5.4xlarge;c5.9xlarge;c5.18xlarge;i3.large;i3.xlarge;i3.2xlarge;i3.4xlarge;i3.8xlarge;i3.16xlarge;r3.xlarge;r3.2xlarge;r3.4xlarge;r3.8xlarge;r4.large;r4.xlarge;r4.2xlarge;r4.4xlarge;r4.8xlarge;r4.16xlarge;x1.16xlarge;x1.32xlarge;p2.xlarge;p2.8xlarge;p2.16xlarge;p3.2xlarge;p3.8xlarge;p3.16xlarge;r5.large;r5.xlarge;r5.2xlarge;r5.4xlarge;r5.12xlarge;r5.24xlarge;r5d.large;r5d.xlarge;r5d.2xlarge;r5d.4xlarge;r5d.12xlarge;r5d.24xlarge;z1d.large;z1d.xlarge;z1d.2xlarge;z1d.3xlarge;z1d.6xlarge;z1d.12xlarge
```
[the remark this answers is not in the document]
Comment (hasheddan) [reply aws/apis/compute/v1alpha1/types.go]: This likely manually added, which is not a sustainable solution. However, it is out of the scope of this PR.

File: azure/apis/storage/v1alpha1/test/account.go [hunk 21]
```diff
@@ -22,18 +22,18 @@ import (
...
-	"github.com/crossplaneio/crossplane/pkg/apis/azure/storage/v1alpha1"
-	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/apis/core/v1alpha1"
+	v1alpha13 "github.com/crossplaneio/crossplane/azure/apis/storage/v1alpha1"
 )
 
 // MockAccount builder for testing account object
 type MockAccount struct {
-	*v1alpha1.Account
+	*v1alpha13.Account
```
[the remark this answers is not in the document]
Comment (hasheddan) [reply azure/apis/storage/v1alpha1/test/account.go]: `Account` has always existed in `azure/storage`. It is not what you would typically think of as an "account". More information [here](https://docs.microsoft.com/en-us/azure/storage/common/storage-account-overview).

File: pkg/controller/gcp/gcp.go (hunk 15 above)
Comment (hasheddan) [reply pkg/controller/gcp/gcp.go]: That is a valid point. I think the logic is that there is no reason to try adding additional controllers when we are certain we will fail. However, this is in-line with how we currently add controllers with the `AddToManager` pattern (return `err` as soon as one is encountered). A change would likely be out of the scope of this PR

File: azure/apis/storage/v1alpha1/test/account.go [hunk 22]
```diff
@@ -22,18 +22,18 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 
-	"github.com/crossplaneio/crossplane/pkg/apis/azure/storage/v1alpha1"
-	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/apis/core/v1alpha1"
+	v1alpha13 "github.com/crossplaneio/crossplane/azure/apis/storage/v1alpha1"
```
[the remark this answers is not in the document]
Comment (hasheddan) [reply azure/apis/storage/v1alpha1/test/account.go]: I agree that would be more consistent with current pattern.

File: cluster/charts/crossplane/crds/aws/cache.aws.crossplane.io_replicationgroups.yaml [hunk 23]
```diff
@@ -89,8 +92,78 @@ spec:
                 you start creating a cluster.
               type: string
             claimRef:
+              description: ObjectReference contains enough information to let you
+                inspect or modify the referred object.
+              properties:
+                apiVersion:
+                  description: API version of the referent.
```
[the remark this answers is not in the document]
Comment (hasheddan) [reply cluster/charts/crossplane/crds/aws/cache.aws.crossplane.io_replicationgroups.yaml]: This is not a change from current, but since you asked: [Referent](https://www.merriam-webster.com/dictionary/referent).

File: pkg/extensions/unpack_test.go (hunk 16 above)
Comment (displague) [reply pkg/extensions/unpack_test.go]: I believe this first came up here: https://github.com/crossplaneio/crossplane/issues/495#issuecomment-495319702

The conditioned status is a list of independent statuses each focused on a particular condition..
```
status:
  conditionedStatus:
    - type: Raining
      status: false
    - type: Sunny
      status: true
```

https://github.com/crossplaneio/crossplane/blob/db468475610f4f83ddc8a1b51e76837f79b6f52d/pkg/apis/core/v1alpha1/condition.go#L56-L97

File: azure/apis/storage/v1alpha1/test/account.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (muvaf) [reply azure/apis/storage/v1alpha1/test/account.go]: That's what GoLand does when moving the file around, one has to go and manually fix them.

**Comment by muvaf** [comment]
Just looked at it one more time and I think we are ready to merge this.

Thanks @jbw976 and @hasheddan for thorough test of each custom resource. I think even if we see some issues here and there after we merge, we do know that our functionality is not broken.
